crates/astria-telemetry/src/metrics/builders.rs

```diff
@@ -0,0 +1,277 @@
+        self.listening_address = Some(listening_address.to_string());
+        self
+    }
+
+    /// Enables or disables setting the global metrics recorder.
+    #[must_use]
+    pub fn with_global_recorder(mut self, use_global_recorder: bool) -> Self {
+        self.use_global_recorder = use_global_recorder;
+        self
+    }
+
+    /// Registers the buckets and metrics as specified in `T::set_buckets` and `T::register`
+    /// respectively, starts the http server if enabled, sets the global metrics recorder if
+    /// requested and returns a new metrics object of type `T` along with a handle for rendering
+    /// current metrics.
+    // allow: no useful error info can be added without writing excessive details.
+    #[allow(clippy::missing_errors_doc)]
+    pub fn build<T: Metrics>(self, config: &T::Config) -> Result<(T, Handle), Error> {
+        // Apply settings to the prometheus builder.
+        let mut prometheus_builder = PrometheusBuilder::new();
+        if !self.service_name.is_empty() {
+            prometheus_builder = prometheus_builder.add_global_label("service", self.service_name);
+        }
+        if let Some(listening_address) = &self.listening_address {
+            let addr: SocketAddr = listening_address.parse()?;
+            prometheus_builder = prometheus_builder.with_http_listener(addr);
+        }
+
+        // Set the histogram buckets.
+        let mut bucket_builder = BucketBuilder {
+            builder: prometheus_builder,
+            buckets: HashMap::new(),
+        };
+        T::set_buckets(&mut bucket_builder, config)?;
+        let histograms_with_buckets = bucket_builder.histogram_names();
+
+        // Consume the prometheus builder, yielding a recorder and a future for running the exporter
+        // server (this will be a no-op if the server isn't configured to run).
+        let (recorder, exporter_fut) = if self.listening_address.is_some() {
+            bucket_builder
+                .builder
+                .build()
+                .map_err(|error| Error::StartListening(error.into()))?
+        } else {
+            let recorder = bucket_builder.builder.build_recorder();
+            let fut: ExporterFuture = Box::pin(async move { Ok(()) });
+            (recorder, fut)
+        };
+        let handle = Handle::new(recorder.handle());
+
+        // Register individual metrics.
+        let mut registering_builder = RegisteringBuilder::new(recorder);
+        let metrics = T::register(&mut registering_builder, config)?;
+
+        // Ensure no histogram buckets were left unassigned.
+        let unassigned: HashSet<_> = histograms_with_buckets
+            .difference(&registering_builder.histograms)
+            .cloned()
+            .collect();
+        if !unassigned.is_empty() {
+            return Err(Error::BucketsNotAssigned(unassigned));
+        }
+
+        // Run the exporter server and set the global recorder if requested.
+        tokio::spawn(exporter_fut);
```

Comment:
I was thinking of maybe not dropping the handle but keeping it around in case something bad happens in the exporter. But it turns out I was thinking of opentelemetry which provides a way to shut down the pipeline. Nothing like that seems to be available for the metrics crate. So keeping the handle is probably useless.